File: Redux+DI/android/app/src/main/java/com/example/weather/WeatherSideEffect.kt
package com.example.weather

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch

class WeatherSideEffect(val weatherApi: WeatherApi) {

    fun fetchWeather(city: String) = GlobalScope.launch {
        val weatherData = weatherApi.getWeather(city)
        GlobalScope.launch(Dispatchers.Main) {
            mainStore.dispatch(ActionGetWeather(weatherData))
        }
    }
}

File: Redux+DI/ios/Weather/WeatherSideEffect.swift
import Foundation
import PromiseKit
import SwiftyJSON
import ReSwift

class WeatherSideEffect {
    let api: WeatherApi
    
    init(api: WeatherApi) {
        self.api = api
    }
    
    func fetchWeather() {
        firstly {
            api.getLocation(city: "Shanghai")
        }.then { city in
            self.api.getWeather(woeid: city["woeid"].intValue)
        }.done { weather in
            mainStore.dispatch(
                ActionGetWeather(weather: weather)
            )
        }.catch {
            mainStore.dispatch(
                ActionError(error: $0)
            )
        }
    }
}
